### PR TITLE
Unification, take 2

### PR DESCRIPTION
**Pull request by traiansf**
Ready for review:  Note that it does not yet solve unification subproblems arising from constraints. I plan to work on that (and additional tests) in the next iteration.

### CODE REVIEW

File: src/main/haskell/language-kore/src/Data/Kore/Unification/Unifier.hs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+-- |'unificationSolutionToPurePattern' packages an unification solution into
+-- a 'CommonPurePattern' by transforming the constraints into a conjunction of
+-- equalities and conjoining them with the unified term.
+unificationSolutionToPurePattern
+    :: MetadataTools level
+    -> UnificationSolution level
+    -> CommonPurePattern level
+unificationSolutionToPurePattern tools ucp =
+    foldr
```
[the remark this answers is not in the document]
Comment (traiansf) [reply src/main/haskell/language-kore/src/Data/Kore/Unification/Unifier.hs]: I wasn't sure what we would really prefer here... My first thoughts were the same as yours, but then I thought one could also prefer having the unification constraint at the top. I'll revise it to your suggested way, nevertheless.

In any case, if we follow the suggestions Xiaohong made in the doc ( https://docs.google.com/document/d/12JudMIqeRfPaewzdyY6LIcjwC4c1UqWRYeyP0X7TyAg/edit#  ) we might not even need to compute the fully combined term as part of the unification process.

File: src/main/haskell/language-kore/src/Data/Kore/Unification/Unifier.hs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    | NonFunctionalHead (SymbolOrAlias level)
+    | NonFunctionalPattern
+    | UnsupportedPatterns
+  deriving (Eq)
+
+-- checks whether a pattern is functional or not
+isFunctionalPattern
+    :: MetadataTools level
+    -> PureMLPattern level Variable
+    -> Either (UnificationError level) [FunctionalProof level]
```
[the remark this answers is not in the document]
Comment (traiansf) [reply src/main/haskell/language-kore/src/Data/Kore/Unification/Unifier.hs]: It is actually the collection of all symbols in a pattern, not just a path.  All of them being functional imply that the pattern itself is functional.

File: src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+solveEqualityConstraint tools (v, p, ecProof) = do
+    (solution, proof) <- simplifyAnd tools p
+    return
+        ( (v, unificationSolutionTerm solution)
+          : unificationSolutionConstraints solution
+        , CombinedUnificationProof [ecProof, proof]
+        )
+
+-- iteratively finds ocurrences of x = t1 /\ x = t2, solves them, and
+-- computes the new substitution
```
Comment (virgil-serbanuta) [inline src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs]: Could you explain (as a code comment) why you can't just split the substitution by variable, then foldr with a combination of normalizeSubstitution1 and solveEqualityConstraint? If you don't already have a test for this case, could you add one?

File: src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+                    )
+            }
+        , AndDistributionAndConstraintLifting
+            (applicationSymbolOrAlias ap)
+            subProofs
+        )
+
+-- finds ocurrences of x = t1 /\ x = t2 and transforms them
+-- into x = (t1 /\ t2)
+normalizeSubstitution1
```
Comment (virgil-serbanuta) [inline src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs]: why not "mergeConsecutiveAssigmenntsForSameVariable" or something similar?

File: src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+    -- ^functions yielding metadata for pattern heads
+    -> CommonPurePattern level
+    -- ^left-hand-side of unification
+    -> CommonPurePattern level
+    -> Either
+        (UnificationError level)
+        (UnificationSubstitution level, UnificationProof level)
+unificationProcedure tools p1 p2
+    | p1Sort /= p2Sort
+    = Left (SortClash p1Sort p2Sort)
```
Comment (virgil-serbanuta) [inline src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs]: Either decrease the indentation by two spaces, or add 4. Or move the '=' on the previous line, I'm not sure.

File: src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+                (extractPurePattern term1)
+                (extractPurePattern term2)
+            )
+        )
+
+unificationTests :: TestTree
+unificationTests =
+    testGroup
+        "Unification Tests"
+        [ unificationSuccess "Constant"
```
Comment (virgil-serbanuta) [inline src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs]: Wouldn't this be more readable as:
```
unificationSuccess "Constant"
    (UnificationTerm aA)
    (UnificationTerm aA)
    (UnificationResult aA)
   ...
```
?

File: src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+unificationSuccess
+    :: (HasCallStack)
+    => String
+    -> CommonPurePatternStub Object
+    -> CommonPurePatternStub Object
+    -> CommonPurePatternStub Object
+    -> Substitution Object
+    -> UnificationProof Object
+    -> TestTree
+unificationSuccess message term1 term2 term subst proof =
```
Comment (virgil-serbanuta) [inline src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs]: Shouldn't `term` be called `result`?

File: src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs (hunk 4 above)
Comment (bmmoore) [reply src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs]: If we're trying to keep to 80-column lines we should not allow identifiers longer than 40 characters!

Comment (traiansf) [reply src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs]: :-)

Comment (virgil-serbanuta) [reply src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs]: We could do that, but why? There are some good arguments for having a limit for line length, and some weaker arguments for it being 80 (or 100) characters. However, I am not familiar with arguments for identifier length (except for one-letter identifiers being bad in many cases).

Anyway, do we have 40-char identifiers in our code? If we have 2-3 examples, we could check if long names help or hurt.

Returning to my comment, the mergeConsecutive... example above was not meant to be copy-pasted into code, I was hoping that Traian would find a better name (that's why I added "or something similar"). Still, I think it's better than normalizeSubstitution1, which, I think, is almost misleading.

File: src/main/haskell/language-kore/src/Data/Kore/ASTPrettyPrint.hs [hunk 8]
```diff
@@ -72,6 +72,32 @@ writeOneFieldStruct
...
+writeTwoFieldStruct flags name contenta contentb =
+    writeOneFieldStructK
+        flags
+        name
+        (  prettyPrint NeedsParentheses contenta
+        >> write " "
+        >> prettyPrint NeedsParentheses contentb
+        )
+
+writeThreeFieldStruct
```
[the remark this answers is not in the document]
Comment (traiansf) [reply src/main/haskell/language-kore/src/Data/Kore/ASTPrettyPrint.hs]: I added a TODO to consider removing them when refactoring the proof objects stuff.

File: src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+                    ]
+                    (var nonLinY)
+                    (extractPurePattern (applyS nonLinG [nonLinX]))
+                , Proposition_5_24_3
+                    [ FunctionalHead (symbolHead nonLinAS) ]
+                    (var nonLinX)
+                    (extractPurePattern nonLinA)
+                ]
+            )
+        , unificationSuccess
```
Comment (virgil-serbanuta) [inline src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs]: s/unificationSuccess/andSimplificationSuccess/?

File: src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+                    ]
+                    (var expX)
+                    (extractPurePattern (applyS expBin [expY, expA]))
+                ]
+            )
+        , unificationProcedureSuccess
+            "times(times(a, y), x) = times(x, times(y, a))"
+            (applyS expBin [applyS expBin [expA, expY], expX])
+            (applyS expBin [expX, applyS expBin [expY, expA]])
+            [ ("a", expY)
```
Comment (virgil-serbanuta) [inline src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs]: Add a todo to solve somehow the a=y, y=a substitution?

File: src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+        , unificationFailure "non-constructor symbolHead left"
+            a2A
+            aA
+            (NonConstructorHead (symbolHead a2))
+        , unificationFailure "nested failure"
+            fa
+            fa1
+            (ConstructorClash (symbolHead a) (symbolHead a1))
+        , unificationFailure "Unsupported constructs"
+            fa
```
Comment (virgil-serbanuta) [inline src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs]: I liked the inline definitions from above more.

File: src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+            faImpliesXa1
+            UnsupportedPatterns
+             {- currently this cannot even be built because of builder checkd
+        , unificationFailure "Unmatching sorts"
+            aA
+            bA
+            UnificationError
+            -}
+        ]
```
Comment (virgil-serbanuta) [inline src/main/haskell/language-kore/test/parser/Data/Kore/Unification/UnifierTest.hs]: How about a test with a unification of f(X, X) with f(a, g(a))?

**Pull request merged**